Quadriga/src/main/java/edu/asu/spring/quadriga/web/config/SimpleSignInAdapter.java

```diff
@@ -0,0 +1,123 @@
+import edu.asu.spring.quadriga.exceptions.UsernameExistsException;
+import edu.asu.spring.quadriga.service.IUserManager;
+import edu.asu.spring.quadriga.service.QuadrigaUserDetails;
+import edu.asu.spring.quadriga.web.config.social.SocialSignInStatus;
+import edu.asu.spring.quadriga.web.login.QuadrigaGrantedAuthority;
+import edu.asu.spring.quadriga.web.manageusers.beans.AccountRequest;
+
+/**
+ * 
+ * This class authorizes (assigns appropriate roles) users authenticated by
+ * Github.
+ * 
+ * @author Chiraag Subramanian
+ *
+ */
+
+public final class SimpleSignInAdapter implements SignInAdapter {
+
+    private final Logger logger = LoggerFactory.getLogger(SignInAdapter.class);
+
+    private IUserManager userManager;
+    private IUserHelper userHelper;
+
+    public SimpleSignInAdapter(IUserManager userManager, IUserHelper userHelper) {
+        this.userManager = userManager;
+        this.userHelper = userHelper;
+    }
+
+    /**
+     * This method checks if Quadriga contains details of the user authenticated
+     * by Github. If the user details are present (which indicates the user is
+     * an authorized user), appropriate roles are assigned, Else if the user
+     * details are not present (which indicates the user is an unregistered /
+     * not approved user), restricted access is assigned.
+     */
+
+    @Override
+    public String signIn(String userId, Connection<?> connection, NativeWebRequest request) {
+
+        List<GrantedAuthority> authorities = new ArrayList<>();
+        SocialSignInStatus messageType = null;
+        IUser user = (IUser) userManager.findUserByProviderUserId(connection.getKey().getProviderUserId(),
+                connection.getKey().getProviderId());
+        // if user details is not present in the database, create a database
+        // record for the user.  
+        if (user == null) {
+            
+            try {
+                    user = userHelper.createUser(userId, connection);
+                    AccountRequest accountRequest = new AccountRequest();
+                    accountRequest.setUsername(user.getUserName());
+                    accountRequest.setName(user.getName());
+                    accountRequest.setEmail(user.getEmail());
+                    accountRequest.setSocialSignIn(true);
+                    accountRequest.setProvider(user.getProvider());
+                    accountRequest.setUserIdOfProvider(user.getUserIdOfProvider());
+                    userManager.addNewUser(accountRequest);
+                    messageType = SocialSignInStatus.REGISTRATION_SUCCESS;
+            } catch (QuadrigaStorageException e){
+                logger.error("Could not add user.", e);
+                messageType = SocialSignInStatus.REGISTRATION_FAILED;
+            } catch(UsernameExistsException e){ 
+                logger.error("User request exists.", e);
+                messageType = SocialSignInStatus.ACCOUNT_APPROVAL_PENDING;
```

Comment:
Doesn't this also mean that the username is actually already in use?

Reply:
Yes. It means that a user request with the <username> was found in the tbl_quadriga_user_requests.

Reply:
actually, it means that either there is a user request pending or that the username already exists. I would change the userManager code to throw different exception depending on the situation so that you can distinguish these two cases, because either the user just have to wait till their account is approved, or someone else choose to use a username that uses the pattern we use for github account usernames (unlikely but possible).